RIOT-OS/RIOT
# makefiles/kconfig: refactor integration and add genconfig script

### Contribution description
This PR refactors the integration of Kconfig into the build system, it adds some improvements and cleanups. This also should fix the issue pointed out by @aabadie [here](https://github.com/RIOT-OS/RIOT/pull/14614#issuecomment-669949800).

#### New `genconfig.py`
The scripts that were taken directly from the kconfiglib repo (`merge_config.py` and `genconfig.py`) are now replaced with a single tool that has both functionalities and is adapted to our build system. The tool allows:
   - merging multiple configuration sources into a single `.config` file
   - generating the `autoconf.h` header file that contains the configurations for compilation
   - generating header files for incremental builds (see #14654)
   - generating a dependency file of the used `Kconfig` files

#### Removal of `merged.config`
This extra file is not used anymore. `menuconfig` edits `out.config` directly, this simplifies the logic and the dependencies in the build system.

#### Introduction of `out.conf.d`
This file is generated via `genconfig.py` and is similar to the `*.d` files generated by GCC. The idea is the same, but based on `Kconfig` files. This allows to re-trigger the generation of `out.conf` whenever a `Kconfig` file that was previously scanned changes.

#### Other changes
- Documentation is updated to the latest integration to the build system
- [`-imacros`](https://gcc.gnu.org/onlinedocs/gcc/Preprocessor-Options.html#Preprocessor-Options) is used to inject `autoconf.h`. This should not change the behaviour but is safer.
- `autoconf.h` is regenerated (if needed) after using `menuconfig`. This helps to keep `out.config` and `autoconf.h` in sync.

### Testing procedure
We should test this carefully. Here I list some workflows that should still be working:

1- On an application *without* `.config` files Kconfig should not be run by default when building. None of the files should be generated.

2- On an application *without* `.config` files Kconfig should run on demand (by setting `SHOULD_RUN_KCONFIG=1` variable in the environment or in the Makefile).
   - The corresponding files should be generated (`bin/<BOARD>/generated`)
   - Configurations changed via `make menuconfig` should re-trigger the generation of `out.config` and `autoconf.h`, and the compilation of the application. New configurations should be applied properly. This should happen *every* time a configuration is changed
   - Cleaning the application without forcing Kconfig should clean all the related files. Moreover, running `make clean all` should not generate any Kconfig file for the new build.

3- On an application *with* `.config` files (e.g. `tests/kconfig`) Kconfig should run by default and its files should be generated. 
   - Configurations should be successfully applied to the application.
   - Changes to the `.config` file should re-trigger building `autoconf.h` and the application. Also, they should be applied to the configuration.
   - Changes via `menuconfig` should have the same effect as the above.

On every case when Kconfig is running, its symbols should be accessible in the build system (e.g. the variables `CONFIG_` are defined).


### Issues/PRs references
Needed for #14654
